File: web/src/services/temporalService.test.ts
import { describe, it, expect } from "vitest";
import {
  getDeviceLocalDate,
  getPlanLocalDate,
  createPlanFromInputs,
  formatPlanDisplay,
  filterTodayTasks,
  filterUpcomingTasks,
  isTaskOverdue,
} from "./temporalService";
import type { Task, Plan } from "../contracts/task";

describe("Temporal Service Seam", () => {
  const baseTask: Task = {
    id: "550e8400-e29b-41d4-a716-446655440000",
    title: "Review deployment",
    description: null,
    priority: 4,
    plan: null,
    labels: [],
    parentId: null,
    completedAt: null,
    createdAt: "2026-08-19T10:00:00.000Z",
    updatedAt: "2026-08-19T10:00:00.000Z",
    version: 1,
  };

  describe("getDeviceLocalDate", () => {
    it("returns YYYY-MM-DD in local time", () => {
      const fixedDate = new Date(2026, 7, 19, 14, 30, 0); // month is 0-indexed: 7 is August
      const localDateStr = getDeviceLocalDate(fixedDate);
      expect(localDateStr).toBe("2026-08-19");
    });
  });

  describe("createPlanFromInputs", () => {
    it("returns null when date is empty or null", () => {
      expect(
        createPlanFromInputs({
          date: null,
          time: null,
          type: null,
          effectiveDefault: "instant",
        }),
      ).toBeNull();

      expect(
        createPlanFromInputs({
          date: "   ",
          time: "10:00",
          type: null,
          effectiveDefault: "instant",
        }),
      ).toBeNull();
    });

    it("returns Date-only plan when date is provided without time (never an Instant, no fake midnight)", () => {
      const plan = createPlanFromInputs({
        date: "2026-08-20",
        time: null,
        type: null,
        effectiveDefault: "instant",
      });

      expect(plan).toEqual({ date: "2026-08-20" });
      expect(plan).not.toHaveProperty("type");
      expect(plan).not.toHaveProperty("at");
      expect(plan).not.toHaveProperty("time");
    });

    it("creates Floating plan when type is explicitly 'floating'", () => {
      const plan = createPlanFromInputs({
        date: "2026-08-20",
        time: "15:30",
        type: "floating",
        effectiveDefault: "instant",
      });

      expect(plan).toEqual({
        type: "floating",
        date: "2026-08-20",
        time: "15:30",
      });
    });

    it("creates Instant plan when type is explicitly 'instant'", () => {
      const plan = createPlanFromInputs({
        date: "2026-08-20",
        time: "15:30",
        type: "instant",
        effectiveDefault: "floating",
      });

      expect(plan?.type).toBe("instant");
      if (plan && plan.type === "instant") {
        expect(plan.at).toMatch(
          /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})$/,
        );
      }
    });

    it("uses effective default ('instant') when no explicit type is provided", () => {
      const plan = createPlanFromInputs({
        date: "2026-08-20",
        time: "09:00",
        type: null,
        effectiveDefault: "instant",
      });

      expect(plan?.type).toBe("instant");
    });

    it("uses effective default ('floating') when no explicit type is provided", () => {
      const plan = createPlanFromInputs({
        date: "2026-08-20",
        time: "09:00",
        type: null,
        effectiveDefault: "floating",
      });

      expect(plan).toEqual({
        type: "floating",
        date: "2026-08-20",
        time: "09:00",
      });
    });

    it("removing clock time produces Date-only plan without fake midnight", () => {
      const plan = createPlanFromInputs({
        date: "2026-08-20",
        time: "",
        type: null,
        effectiveDefault: "instant",
      });

      expect(plan).toEqual({ date: "2026-08-20" });
    });
  });

  describe("getPlanLocalDate", () => {
    it("returns null for null plan", () => {
      expect(getPlanLocalDate(null)).toBeNull();
    });

    it("returns exact date for Date-only plan", () => {
      expect(getPlanLocalDate({ date: "2026-08-20" })).toBe("2026-08-20");
    });

    it("returns exact date for Floating plan", () => {
      expect(
        getPlanLocalDate({
          type: "floating",
          date: "2026-08-20",
          time: "10:00",
        }),
      ).toBe("2026-08-20");
    });

    it("returns device-local date for Instant plan UTC timestamp", () => {
      const plan: Plan = {
        type: "instant",
        at: "2026-08-19T23:30:00.000Z",
      };
      const localDate = getPlanLocalDate(plan);
      expect(localDate).toMatch(/^\d{4}-\d{2}-\d{2}$/);
    });
  });

  describe("filterTodayTasks", () => {
    it("returns open tasks whose device-local plan day is today or earlier", () => {
      const todayDate = "2026-08-19";
      const now = new Date(2026, 7, 19, 12, 0, 0);

      const todayTask: Task = {
        ...baseTask,
        id: "11111111-1111-1111-1111-111111111111",
        plan: { date: todayDate },
      };
      const overdueTask: Task = {
        ...baseTask,
        id: "22222222-2222-2222-2222-222222222222",
        plan: { date: "2026-08-18" },
      };
      const futureTask: Task = {
        ...baseTask,
        id: "33333333-3333-3333-3333-333333333333",
        plan: { date: "2026-08-20" },
      };
      const undatedTask: Task = {
        ...baseTask,
        id: "44444444-4444-4444-4444-444444444444",
        plan: null,
      };
      const completedTodayTask: Task = {
        ...baseTask,
        id: "55555555-5555-5555-5555-555555555555",
        plan: { date: todayDate },
        completedAt: "2026-08-19T11:00:00Z",
      };

      const result = filterTodayTasks(
        [todayTask, overdueTask, futureTask, undatedTask, completedTodayTask],
        now,
      );

      expect(result.map((t) => t.id)).toEqual([overdueTask.id, todayTask.id]);
    });
  });

  describe("filterUpcomingTasks", () => {
    it("groups open dated tasks from today forward and separates overdue tasks with no 7-day cap", () => {
      const now = new Date(2026, 7, 19, 12, 0, 0);

      const overdueTask: Task = {
        ...baseTask,
        id: "22222222-2222-2222-2222-222222222222",
        title: "Overdue task",
        plan: { date: "2026-08-17" },
      };
      const todayTask: Task = {
        ...baseTask,
        id: "11111111-1111-1111-1111-111111111111",
        title: "Today task",
        plan: { date: "2026-08-19" },
      };
      const tomorrowTask: Task = {
        ...baseTask,
        id: "33333333-3333-3333-3333-333333333333",
        title: "Tomorrow task",
        plan: { date: "2026-08-20" },
      };
      const farFutureTask: Task = {
        ...baseTask,
        id: "66666666-6666-6666-6666-666666666666",
        title: "Far future task (30 days)",
        plan: { date: "2026-09-19" },
      };
      const undatedTask: Task = {
        ...baseTask,
        id: "44444444-4444-4444-4444-444444444444",
        plan: null,
      };
      const completedTask: Task = {
        ...baseTask,
        id: "55555555-5555-5555-5555-555555555555",
        plan: { date: "2026-08-20" },
        completedAt: "2026-08-19T10:00:00Z",
      };

      const result = filterUpcomingTasks(
        [
          farFutureTask,
          todayTask,
          overdueTask,
          tomorrowTask,
          undatedTask,
          completedTask,
        ],
        now,
      );

      expect(result.overdue.map((t) => t.id)).toEqual([overdueTask.id]);
      expect(result.groups).toHaveLength(3);
      expect(result.groups[0].date).toBe("2026-08-19");
      expect(result.groups[0].tasks.map((t) => t.id)).toEqual([todayTask.id]);
      expect(result.groups[1].date).toBe("2026-08-20");
      expect(result.groups[1].tasks.map((t) => t.id)).toEqual([
        tomorrowTask.id,
      ]);
      expect(result.groups[2].date).toBe("2026-09-19");
      expect(result.groups[2].tasks.map((t) => t.id)).toEqual([
        farFutureTask.id,
      ]);
    });
  });

  describe("formatPlanDisplay", () => {
    it("formats Date-only plan correctly", () => {
      const now = new Date(2026, 7, 19, 12, 0, 0);
      const display = formatPlanDisplay({ date: "2026-08-19" }, { now });

      expect(display).toEqual({
        dateLabel: "Today",
        timeLabel: null,
        typeLabel: null,
        isOverdue: false,
      });
    });

    it("formats Instant plan deriving device local time and labeling type", () => {
      const instantMoment = new Date(2026, 7, 20, 14, 45, 0); // Local 14:45
      const now = new Date(2026, 7, 19, 12, 0, 0);
      const display = formatPlanDisplay(
        { type: "instant", at: instantMoment.toISOString() },
        { now },
      );

      expect(display?.dateLabel).toBe("Tomorrow");
      expect(display?.timeLabel).toBe("14:45");
      expect(display?.typeLabel).toBe("Instant");
      expect(display?.isOverdue).toBe(false);
    });

    it("flags overdue tasks appropriately", () => {
      const now = new Date(2026, 7, 19, 12, 0, 0);
      const display = formatPlanDisplay({ date: "2026-08-18" }, { now });

      expect(display?.isOverdue).toBe(true);
      expect(display?.dateLabel).toBe("Yesterday");
    });
  });

  describe("Controlled-Clock Edge Cases", () => {
    it("resolves relative date immediately at midnight boundary", () => {
      const beforeMidnight = new Date(2026, 7, 19, 23, 59, 59);
      const afterMidnight = new Date(2026, 7, 20, 0, 0, 1);

      expect(getDeviceLocalDate(beforeMidnight)).toBe("2026-08-19");
      expect(getDeviceLocalDate(afterMidnight)).toBe("2026-08-20");

      const task: Task = {
        ...baseTask,
        plan: { date: "2026-08-19" },
      };

      // Before midnight, task is today (not overdue)
      expect(isTaskOverdue(task, beforeMidnight)).toBe(false);
      // After midnight, task becomes overdue
      expect(isTaskOverdue(task, afterMidnight)).toBe(true);
    });

    it("resolves Instant plan date according to viewing device local date", () => {
      // 2026-08-19T22:30:00.000Z
      const utcIso = "2026-08-19T22:30:00.000Z";
      const instantPlan: Plan = {
        type: "instant",
        at: utcIso,
      };

      const dateObj = new Date(utcIso);
      const expectedLocalDate = getDeviceLocalDate(dateObj);
      expect(getPlanLocalDate(instantPlan)).toBe(expectedLocalDate);
    });

    it("preserves Floating date and time regardless of timezone or DST", () => {
      const floatingPlan: Plan = {
        type: "floating",
        date: "2026-03-29",
        time: "02:30",
      };

      expect(getPlanLocalDate(floatingPlan)).toBe("2026-03-29");
      const display = formatPlanDisplay(floatingPlan, {
        now: new Date(2026, 2, 28, 12, 0, 0),
      });
      expect(display?.timeLabel).toBe("02:30");
      expect(display?.typeLabel).toBe("Floating");
    });
  });
});

File: android/app/src/test/java/com/cras/app/domain/TemporalServiceTest.kt
package com.cras.app.domain

import com.cras.app.models.Plan
import com.cras.app.models.Task
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.ZoneOffset
import java.util.UUID

class TemporalServiceTest {

    private val tokyoZone = ZoneId.of("Asia/Tokyo") // UTC+9
    private val nyZone = ZoneId.of("America/New_York") // UTC-4 in summer
    private val utcZone = ZoneOffset.UTC

    private fun createTask(
        title: String,
        plan: Plan? = null,
        priority: Int = 4,
        completedAt: String? = null
    ): Task = Task(
        id = UUID.randomUUID().toString(),
        title = title,
        description = null,
        priority = priority,
        plan = plan,
        labels = emptyList(),
        parentId = null,
        completedAt = completedAt,
        createdAt = "2026-08-19T00:00:00Z",
        updatedAt = "2026-08-19T00:00:00Z",
        version = 1
    )

    @Test
    fun `getDeviceLocalDate returns local calendar date YYYY-MM-DD for given instant and timezone`() {
        // 2026-08-19T20:00:00Z is 2026-08-19 in UTC, 2026-08-19 in New York (16:00), but 2026-08-20 in Tokyo (05:00 next day)
        val instant = Instant.parse("2026-08-19T20:00:00Z")

        assertEquals("2026-08-19", getDeviceLocalDate(instant, utcZone))
        assertEquals("2026-08-19", getDeviceLocalDate(instant, nyZone))
        assertEquals("2026-08-20", getDeviceLocalDate(instant, tokyoZone))
    }

    @Test
    fun `getPlanLocalDate extracts local calendar date for all plan variants`() {
        // Date-only
        val dateOnly = Plan.DateOnly("2026-08-19")
        assertEquals("2026-08-19", getPlanLocalDate(dateOnly, utcZone))
        assertEquals("2026-08-19", getPlanLocalDate(dateOnly, tokyoZone))

        // Floating: identical date across all timezones
        val floating = Plan.Floating("2026-08-19", "09:30")
        assertEquals("2026-08-19", getPlanLocalDate(floating, utcZone))
        assertEquals("2026-08-19", getPlanLocalDate(floating, tokyoZone))

        // Instant: derived from viewing device local timezone
        val instantPlan = Plan.Instant("2026-08-19T22:00:00Z")
        assertEquals("2026-08-19", getPlanLocalDate(instantPlan, utcZone))
        assertEquals("2026-08-19", getPlanLocalDate(instantPlan, nyZone))
        assertEquals("2026-08-20", getPlanLocalDate(instantPlan, tokyoZone))

        // Null plan
        assertNull(getPlanLocalDate(null, utcZone))
    }

    @Test
    fun `createPlanFromInputs creates contract-compliant Date-only, Floating, and Instant plans`() {
        // 1. Omitted date -> null
        assertNull(createPlanFromInputs(CreatePlanParams(date = null)))
        assertNull(createPlanFromInputs(CreatePlanParams(date = "   ")))

        // 2. Date provided, time omitted -> DateOnly (no timed type, no fake midnight)
        val dateOnly = createPlanFromInputs(CreatePlanParams(date = "2026-08-19"))
        assertTrue("Expected DateOnly plan", dateOnly is Plan.DateOnly)
        assertEquals("2026-08-19", (dateOnly as Plan.DateOnly).date)

        // 3. Date & time provided with explicit Floating
        val floatingPlan = createPlanFromInputs(
            CreatePlanParams(
                date = "2026-08-19",
                time = "14:30",
                type = TimedPlanType.FLOATING
            )
        )
        assertTrue("Expected Floating plan", floatingPlan is Plan.Floating)
        val floating = floatingPlan as Plan.Floating
        assertEquals("2026-08-19", floating.date)
        assertEquals("14:30", floating.time)

        // 4. Date & time provided with explicit Instant in specific timezone
        val instantPlan = createPlanFromInputs(
            CreatePlanParams(
                date = "2026-08-19",
                time = "14:30",
                type = TimedPlanType.INSTANT,
                zoneId = nyZone // UTC-4 in August -> 14:30 EDT = 18:30 UTC
            )
        )
        assertTrue("Expected Instant plan", instantPlan is Plan.Instant)
        val instant = instantPlan as Plan.Instant
        assertEquals("2026-08-19T18:30:00Z", instant.at)

        // 5. Inherits effective default when type is null
        val defaultFloatingPlan = createPlanFromInputs(
            CreatePlanParams(
                date = "2026-08-19",
                time = "09:00",
                type = null,
                effectiveDefault = TimedPlanType.FLOATING
            )
        )
        assertTrue(defaultFloatingPlan is Plan.Floating)

        val defaultInstantPlan = createPlanFromInputs(
            CreatePlanParams(
                date = "2026-08-19",
                time = "09:00",
                type = null,
                effectiveDefault = TimedPlanType.INSTANT,
                zoneId = utcZone
            )
        )
        assertTrue(defaultInstantPlan is Plan.Instant)
        assertEquals("2026-08-19T09:00:00Z", (defaultInstantPlan as Plan.Instant).at)
    }

    @Test
    fun `formatFriendlyDateLabel formats Today, Tomorrow, Yesterday and dates correctly`() {
        assertEquals("Today", formatFriendlyDateLabel("2026-08-19", "2026-08-19", "2026-08-20", "2026-08-18"))
        assertEquals("Tomorrow", formatFriendlyDateLabel("2026-08-20", "2026-08-19", "2026-08-20", "2026-08-18"))
        assertEquals("Yesterday", formatFriendlyDateLabel("2026-08-18", "2026-08-19", "2026-08-20", "2026-08-18"))
        assertEquals("Fri, Aug 21", formatFriendlyDateLabel("2026-08-21", "2026-08-19", "2026-08-20", "2026-08-18"))
    }

    @Test
    fun `formatPlanDisplay formats plan display info with date, time, type and overdue status`() {
        val now = Instant.parse("2026-08-19T12:00:00Z")

        // Overdue date-only task
        val overduePlan = Plan.DateOnly("2026-08-18")
        val overdueDisplay = formatPlanDisplay(overduePlan, now, utcZone)
        assertNotNull(overdueDisplay)
        assertEquals("Yesterday", overdueDisplay!!.dateLabel)
        assertNull(overdueDisplay.timeLabel)
        assertNull(overdueDisplay.typeLabel)
        assertTrue(overdueDisplay.isOverdue)

        // Today floating task
        val todayFloating = Plan.Floating("2026-08-19", "15:00")
        val todayDisplay = formatPlanDisplay(todayFloating, now, utcZone)
        assertNotNull(todayDisplay)
        assertEquals("Today", todayDisplay!!.dateLabel)
        assertEquals("15:00", todayDisplay.timeLabel)
        assertEquals("Floating", todayDisplay.typeLabel)
        assertFalse(todayDisplay.isOverdue)

        // Tomorrow instant task in NY timezone
        // 2026-08-20T19:00:00Z is 15:00 EDT in New York
        val instantPlan = Plan.Instant("2026-08-20T19:00:00Z")
        val instantDisplay = formatPlanDisplay(instantPlan, now, nyZone)
        assertNotNull(instantDisplay)
        assertEquals("Tomorrow", instantDisplay!!.dateLabel)
        assertEquals("15:00", instantDisplay.timeLabel)
        assertEquals("Instant", instantDisplay.typeLabel)
        assertFalse(instantDisplay.isOverdue)
    }

    @Test
    fun `filterTodayTasks includes open tasks scheduled today or earlier and excludes completed or undated tasks`() {
        val now = Instant.parse("2026-08-19T12:00:00Z") // 2026-08-19 in UTC

        val overdueTask = createTask("Overdue P2", Plan.DateOnly("2026-08-18"), priority = 2)
        val todayP1Task = createTask("Today P1", Plan.Floating("2026-08-19", "10:00"), priority = 1)
        val todayP3Task = createTask("Today P3", Plan.Instant("2026-08-19T14:00:00Z"), priority = 3)
        val tomorrowTask = createTask("Tomorrow", Plan.DateOnly("2026-08-20"), priority = 1)
        val completedTodayTask = createTask("Completed Today", Plan.DateOnly("2026-08-19"), completedAt = "2026-08-19T10:00:00Z")
        val inboxTask = createTask("Inbox task", plan = null)

        val allTasks = listOf(tomorrowTask, inboxTask, todayP3Task, overdueTask, completedTodayTask, todayP1Task)

        val todayFiltered = filterTodayTasks(allTasks, now, utcZone)

        assertEquals(3, todayFiltered.size)
        // Ordered by date ascending, then priority
        assertEquals("Overdue P2", todayFiltered[0].title)
        assertEquals("Today P1", todayFiltered[1].title)
        assertEquals("Today P3", todayFiltered[2].title)
    }

    @Test
    fun `filterUpcomingTasks groups open future dated tasks and separates overdue strip`() {
        val now = Instant.parse("2026-08-19T12:00:00Z") // 2026-08-19

        val overdue1 = createTask("Overdue older", Plan.DateOnly("2026-08-17"), priority = 3)
        val overdue2 = createTask("Overdue yesterday", Plan.DateOnly("2026-08-18"), priority = 1)
        val todayTask = createTask("Today task", Plan.Floating("2026-08-19", "10:00"), priority = 2)
        val tomorrowP2 = createTask("Tomorrow P2", Plan.DateOnly("2026-08-20"), priority = 2)
        val tomorrowP1 = createTask("Tomorrow P1", Plan.DateOnly("2026-08-20"), priority = 1)
        val nextWeek = createTask("Next week", Plan.DateOnly("2026-08-26"), priority = 4)
        val completedTask = createTask("Completed future", Plan.DateOnly("2026-08-21"), completedAt = "2026-08-19T00:00:00Z")
        val undatedTask = createTask("Undated inbox", plan = null)

        val allTasks = listOf(nextWeek, undatedTask, tomorrowP2, overdue2, completedTask, todayTask, overdue1, tomorrowP1)

        val result = filterUpcomingTasks(allTasks, now, utcZone)

        // Overdue strip contains tasks before today sorted date asc
        assertEquals(2, result.overdue.size)
        assertEquals("Overdue older", result.overdue[0].title)
        assertEquals("Overdue yesterday", result.overdue[1].title)

        // Day groups: 2026-08-19, 2026-08-20, 2026-08-26
        assertEquals(3, result.groups.size)

        val todayGroup = result.groups[0]
        assertEquals("2026-08-19", todayGroup.date)
        assertEquals("Today", todayGroup.dateLabel)
        assertEquals(1, todayGroup.tasks.size)
        assertEquals("Today task", todayGroup.tasks[0].title)

        val tomorrowGroup = result.groups[1]
        assertEquals("2026-08-20", tomorrowGroup.date)
        assertEquals("Tomorrow", tomorrowGroup.dateLabel)
        assertEquals(2, tomorrowGroup.tasks.size)
        assertEquals("Tomorrow P1", tomorrowGroup.tasks[0].title)
        assertEquals("Tomorrow P2", tomorrowGroup.tasks[1].title)

        val nextWeekGroup = result.groups[2]
        assertEquals("2026-08-26", nextWeekGroup.date)
        assertEquals(1, nextWeekGroup.tasks.size)
        assertEquals("Next week", nextWeekGroup.tasks[0].title)
    }
}
